clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -66,17 +69,102 @@
+    private static class TopicPartitionBookkeeper {
 
-    // The sequence of the last record of the last ack'd batch from the given partition. When there are no
-    // in flight requests for a partition, the lastAckedSequence(topicPartition) == nextSequence(topicPartition) - 1.
-    private final Map<TopicPartition, Integer> lastAckedSequence;
+        private final Map<TopicPartition, TopicPartitionEntry> topicPartitionBookkeeping = new HashMap<>();
+
+        boolean hasPendingOffsetCommits() {
+            return topicPartitionBookkeeping.values().stream().anyMatch(tpe -> tpe.pendingTxnOffsetCommit != null);
+        }
+
+        void clearPendingOffsetCommits() {
+            topicPartitionBookkeeping.values().forEach(tpe -> tpe.pendingTxnOffsetCommit = null);
+        }
+
+        void clearPendingOffsetCommit(TopicPartition topic) {
+            TopicPartitionEntry tpe = topicPartitionBookkeeping.get(topic);
+            if (tpe != null) {
+                tpe.pendingTxnOffsetCommit = null;
+            }
+        }
+
+        Map<TopicPartition, CommittedOffset> pendingOffsetCommits() {
+            return topicPartitionBookkeeping.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue() != null && e.getValue().pendingTxnOffsetCommit != null)
+                    .collect(Collectors.toMap(Map.Entry::getKey, t -> t.getValue().pendingTxnOffsetCommit));
+        }
+
+        public TopicPartitionEntry get(TopicPartition topic) {
+            TopicPartitionEntry ent = topicPartitionBookkeeping.get(topic);
+            if (ent == null) {
+                ent = new TopicPartitionEntry();
+                topicPartitionBookkeeping.put(topic, ent);
+            }
+            return ent;
+        }
+
+        boolean contains(TopicPartition partition) {
+            return topicPartitionBookkeeping.containsKey(partition);
+        }
+
+        public void reset() {
+            topicPartitionBookkeeping.clear();
+        }
+
+        OptionalLong getLastAckedOffset(TopicPartition partition) {
+            TopicPartitionEntry entry = topicPartitionBookkeeping.get(partition);
+            if (entry != null && entry.lastAckedOffset != ProduceResponse.INVALID_OFFSET)
+                return OptionalLong.of(entry.lastAckedOffset);
+            else
+                return OptionalLong.empty();
+        }
+
+        OptionalInt getLastAckedSequenceNumber(TopicPartition partition) {
+            TopicPartitionEntry entry = topicPartitionBookkeeping.get(partition);
+            if (entry != null && entry.lastAckedSequenceNumber != NO_LAST_ACKED_SEQUENCE_NUMBER)
+                return OptionalInt.of(entry.lastAckedSequenceNumber);
+            else
+                return OptionalInt.empty();
+        }
+    }
+
+    private static class TopicPartitionEntry {
+
+        // The base sequence of the next batch bound for a given partition.
+        private int nextSequenceNumber;
```

Comment:
We could, although calling it `sequenceNumber` was a better term for me at the beginning. By `sequence` I always expected to get or see a list of some kind.

Reply:
Renamed them to *Sequence :)

Reply:
Yeah, it's a bit subjective. I find the (over?)verbosity irritating since I have done a lot of work in this code even if the name is technically more accurate. In any case, I think we'd agree that we should try to be consistent.